### PR TITLE
feature: store metrics in CnosDB

### PR DESCRIPTION
**Pull request by ZuoTiJia**
Now there are 5 metrics that can be stored in db 'usage_schema
- queries
- data_in
- data_out
- writes
- TsFamily disk_storage(not include wal)

### CODE REVIEW

**Review by roseboy-liu — COMMENTED** [review]
add basic test case for your feature

File: query_server/query/src/metadata/usage_schema_provider/mod.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+pub fn create_usage_schema_view_table(
+    user: &User,
+    coord: CoordinatorRef,
+    meta: MetaClientRef,
+    view_table_name: &str,
+) -> spi::Result<Arc<dyn TableProvider>> {
+    let tskv = coord
+        .tenant_meta("cnosdb")
```
Comment (yukkit) [inline query_server/query/src/metadata/usage_schema_provider/mod.rs:94]: use global val

File: coordinator/src/service.rs [hunk 2]
```diff
@@ -315,24 +366,21 @@ impl CoordService {
             let _ = sender.send(Err(e)).await;
             return;
         }
+        let executor = QueryExecutor::new(option, kv_inst, meta, sender.clone(), metrics);
```
Comment (yukkit) [inline coordinator/src/service.rs:369]: could use `QueryExecutorFactory` instead, so you don't need to pass in `metrics: Arc<MetricsRegister>` here

File: coordinator/src/service.rs [hunk 3]
```diff
@@ -378,6 +426,70 @@ impl CoordService {
...
+    //                 ],
+    //                 None,
+    //             );
+    //             let metric = Metric::from_counter(des, writes_counter.clone());
+    //             res.push(metric)
+    //         }
+    //     }
+    //     res
+    // }
+    // }
```
Comment (yukkit) [inline coordinator/src/service.rs:492]: remove

File: coordinator/src/service.rs [hunk 4]
```diff
@@ -406,9 +518,19 @@ impl Coordinator for CoordService {
...
-        let data_len = request.points.len();
-        limiter.check_data_in(data_len).await?;
+        limiter.check_data_in(write_size).await?;
+
+        let db = get_db_from_flatbuffers(points)?;
+        let metrics = self
+            .metrics
+            .root_register()
+            .register_metric::<U64Counter>("data_in", "tenant data in");
+        let counter = metrics.recorder([("tenant", tenant.as_str()), ("database", db.as_str())]);
```
Comment (yukkit) [inline coordinator/src/service.rs:532]: init with coord

File: main/src/tcp/tcp_service.rs [hunk 5]
```diff
@@ -345,31 +374,46 @@ async fn query_record_batch(
...
     for id in cmd.args.vnode_ids.iter() {
         vnodes.push(VnodeInfo { id: *id, node_id })
     }
 
-    let executor = QueryExecutor::new(option, Some(kv_inst), meta, sender.clone());
+    // let db_metrics = coord
+    //     .metrics()
+    //     .tenant_metrics(option.tenant.clone())
+    //     .db_metrics(option.table_schema.db.clone());
+
```
Comment (yukkit) [inline main/src/tcp/tcp_service.rs:409]: remove

File: main/src/tcp/tcp_service.rs [hunk 6]
```diff
@@ -345,31 +374,46 @@ async fn query_record_batch(
...
+            cmd.args.batch_size,
+            cmd.args.tenant.clone(),
+            filter,
+            cmd.expr.df_schema,
+            cmd.expr.table_schema,
+            scan_metrics.tskv_metrics(),
+        )
+    };
 
+    // let meta = coord.meta_manager();
```
Comment (yukkit) [inline main/src/tcp/tcp_service.rs:398]: remove

File: query_server/query/src/instance.rs [hunk 7]
```diff
@@ -30,6 +31,7 @@ pub struct Cnosdbms<D> {
     access_control: AccessControlRef,
     // query dispatcher & query execution
     query_dispatcher: D,
+    // metrics_register: Arc<MetricsRegister>,
```
Comment (yukkit) [inline query_server/query/src/instance.rs:34]: remove

File: query_server/query/src/instance.rs [hunk 8]
```diff
@@ -124,6 +126,7 @@ pub async fn make_cnosdbms(
 
     let db_server = builder
         .query_dispatcher(query_dispatcher)
+        // .metrics_register(metrics_register)
```
Comment (yukkit) [inline query_server/query/src/instance.rs:129]: remove

File: coordinator/src/service.rs [hunk 9]
```diff
@@ -172,6 +189,39 @@ impl CoordService {
...
+        coord: Arc<CoordService>,
+        root_metrics_register: Arc<MetricsRegister>,
+    ) {
+        let start = tokio::time::Instant::now() + tokio::time::Duration::from_secs(10);
+        let interval = tokio::time::Duration::from_secs(10);
+        let mut intv = tokio::time::interval_at(start, interval);
+        loop {
+            intv.tick().await;
+            if let Some(engine) = coord.kv_inst.as_ref() {
+                engine.disk_storage().await;
```
Comment (yukkit) [inline coordinator/src/service.rs:202]: what happen

File: tskv/src/database.rs [hunk 10]
```diff
@@ -35,6 +39,12 @@ use crate::{ColumnFileId, TimeRange, TseriesFamilyId};
 
 pub type FlatBufferPoint<'a> = flatbuffers::Vector<'a, flatbuffers::ForwardsUOffset<Point<'a>>>;
 
+pub struct TSKVDBMetrics {}
+
+pub struct TSKVIndexMetrics {}
+
+pub struct TseriesFamilyMetrics {}
```
Comment (yukkit) [inline tskv/src/database.rs:46]: what

File: tskv/src/database.rs [hunk 11]
```diff
@@ -122,6 +136,7 @@ impl Database {
             LevelInfo::init_levels(self.owner.clone(), tsf_id, self.opt.storage.clone()),
             i64::MIN,
             Arc::new(ShardedCache::default()),
+            Arc::new(MetricsRegister::default()),
```
Comment (yukkit) [inline tskv/src/database.rs:139]: Can `default` be used here?

File: tskv/src/engine.rs [hunk 12]
```diff
@@ -120,6 +121,8 @@ pub trait Engine: Send + Sync + Debug {
     async fn drop_vnode(&self, id: TseriesFamilyId) -> Result<()>;
 
     async fn compact(&self, tenant: &str, database: &str);
+
+    async fn disk_storage(&self);
```
Comment (yukkit) [inline tskv/src/engine.rs:125]: I don't understand the purpose here

File: common/metrics/Cargo.toml [hunk 13]
```diff
@@ -7,5 +7,9 @@ edition.workspace = true
 
 [dependencies]
 trace = { path = "../trace" }
+protos = { path = "../protos", features = ["test"] }
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply common/metrics/Cargo.toml]: already remove

File: coordinator/src/service.rs [hunk 14]
```diff
@@ -206,6 +223,34 @@ impl CoordService {
         }
     }
 
+    async fn metrics_service(coord: Arc<CoordService>) {
+        loop {
+            let start = time::Instant::now() + time::Duration::from_secs(5);
+            let interval = time::Duration::from_secs(10);
+            let mut intv = time::interval_at(start, interval);
+            intv.tick().await;
+            let mut metrics = Coordinator::metrics(coord.as_ref()).await;
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply coordinator/src/service.rs]: already remove

File: coordinator/src/service.rs [hunk 15]
```diff
@@ -233,35 +278,55 @@ impl CoordService {
...
+        let db_info =
+            meta_client
+                .get_db_info(&db_name)?
+                .ok_or_else(|| MetaError::DatabaseNotFound {
+                    database: db_name.to_string(),
+                })?;
+
+        // points_len * (replica number - 1)
+        let data_len =
+            req.request.points.len() * (db_info.schema.config.replica_or_default() - 1) as usize;
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply coordinator/src/service.rs]: already refactor

File: coordinator/src/service.rs [hunk 16]
```diff
@@ -365,31 +430,40 @@ impl CoordService {
         self.exec_on_node(req_node_id, cmd).await
     }
 
-    async fn select_statement_request(coord: Arc<CoordService>, req: SelectStatementRequest) {
+    async fn select_statement_request_handle(
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply coordinator/src/service.rs]: already remove

File: coordinator/src/writer.rs [hunk 17]
```diff
@@ -128,6 +131,11 @@ impl<'a> VnodeMapping<'a> {
                 });
             }
         }
+        let db_info = meta_client.get_db_info(point.db.as_str())?.ok_or_else(|| {
+            MetaError::DatabaseNotFound {
+                database: point.db.to_string(),
+            }
+        })?;
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply coordinator/src/writer.rs]: already remove

File: coordinator/src/writer.rs [hunk 18]
```diff
@@ -197,15 +209,27 @@ impl PointWriter {
             })?;
 
             mapping.map_point(meta_client.clone(), point)?;
+            coord.metrics();
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply coordinator/src/writer.rs]: already removed

File: query_server/query/src/metadata/usage_schema_provider/mod.rs [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+        .ok_or_else(|| MetaError::TenantNotFound {
+            tenant: DEFAULT_CATALOG.into(),
+        })?
+        .get_tskv_table_schema(USAGE_SCHEMA, view_table_name)?
+        .ok_or_else(|| MetaError::TableNotFound {
+            table: view_table_name.into(),
+        })?;
+    let cluster_table = Arc::new(ClusterTable::new(coord.clone(), tskv));
+    if user.desc().name().eq(ROOT) && meta.tenant_name().eq(DEFAULT_CATALOG) {
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply query_server/query/src/metadata/usage_schema_provider/mod.rs]: already refactor

File: common/metrics/src/metric_system.rs [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+pub trait MetricSource: Debug {
+    fn name(&self) -> Option<&'static str>;
+    fn labels(&self) -> Labels;
+    fn register(&self) -> Arc<MetricsRegister>;
+}
+
+// MetricsSystem， 全局唯一 ::new 有一个参数是配置文件，配置文件里面的内容是采集metrics的reporter，根据配置文件，生成各种reporter
+// reporter 持有 Arc<register>, 当 reporter
+//
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply common/metrics/src/metric_system.rs]: already remove

File: coordinator/src/reader.rs [hunk 21]
```diff
@@ -190,6 +196,14 @@ impl QueryExecutor {
                         .await
                         .check_data_out(rsp.record.get_array_memory_size())
                         .await?;
+                    let metrics = self
+                        .metrics_register
+                        .register_metric::<U64Counter>("data_out", "tenant data out statistic");
+                    let counter = metrics.recorder([
+                        ("tenant", self.option.tenant.as_str()),
+                        ("database", self.option.table_schema.db.as_str()),
+                    ]);
```
[the remark this answers is not in the document]
Comment (ZuoTiJia) [reply coordinator/src/reader.rs]: already refactor

File: query_server/query/src/instance.rs (hunk 8 above)
Comment (ZuoTiJia) [reply query_server/query/src/instance.rs]: already remove

File: coordinator/src/service.rs (hunk 9 above)
Comment (ZuoTiJia) [reply coordinator/src/service.rs]: already refactor

File: tskv/src/database.rs (hunk 10 above)
Comment (ZuoTiJia) [reply tskv/src/database.rs]: already remove

File: tskv/src/database.rs (hunk 11 above)
Comment (ZuoTiJia) [reply tskv/src/database.rs]: already fix

File: tskv/src/engine.rs (hunk 12 above)
Comment (ZuoTiJia) [reply tskv/src/engine.rs]: already fix

File: main/src/http/metrics.rs [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+use metrics::metric::Metric;
+use metrics::metric_register::MetricsRegister;
+
+pub struct HttpMetrics {
+    queries: Metric<U64Counter>,
+    writes: Metric<U64Counter>,
+}
+
+impl HttpMetrics {
+    pub fn new(register: Arc<MetricsRegister>) -> Self {
```
Comment (yukkit) [inline main/src/http/metrics.rs:14]: Suggested pass by `&Arc<MetricsRegister>`，because you don't hold it

File: tskv/src/version_set.rs [hunk 23]
```diff
@@ -60,8 +70,20 @@ impl VersionSet {
                     Some(schema) => schema,
                 },
             };
+
+            let register = metrics_register.sub_register([
```
Comment (yukkit) [inline tskv/src/version_set.rs:74]: I didn't find where to use it

File: tskv/src/summary.rs [hunk 24]
```diff
@@ -463,6 +475,7 @@ impl Summary {
                 levels,
                 max_level_ts,
                 Arc::new(ShardedCache::default()),
+                Arc::new(MetricsRegister::default()),
```
Comment (yukkit) [inline tskv/src/summary.rs:478]: Do you really want to use default here?

File: query_server/query/src/metadata/usage_schema_provider/mod.rs [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+    ) -> Result<Arc<dyn TableProvider>>;
+}
+
+pub fn create_usage_schema_view_table(
+    user: &User,
+    coord: CoordinatorRef,
+    meta: MetaClientRef,
+    view_table_name: &str,
+) -> spi::Result<Arc<dyn TableProvider>> {
+    let tskv = block_on(coord.tenant_meta(DEFAULT_CATALOG))
```
Comment (yukkit) [inline query_server/query/src/metadata/usage_schema_provider/mod.rs:92]: Is it possible to hold `coord.tenant_meta(DEFAULT_CATALOG)` at initialization

File: main/src/main.rs [hunk 26]
```diff
@@ -88,7 +90,7 @@ fn main() -> Result<(), std::io::Error> {
     let cli = Cli::parse();
     let config = parse_config(&cli);
 
-    let _ = init_global_tracing(&config.log.path, "tsdb.log", &config.log.level);
+    let _ = init_global_tracing(&config.log.path, "tsdb.log", &config.log.level, 6669);
```
Comment (yukkit) [inline main/src/main.rs:93]: hardcode

File: tskv/src/tseries_family.rs [hunk 27]
```diff
@@ -601,6 +610,30 @@ impl SuperVersion {
     }
 }
 
+#[derive(Debug)]
+pub struct TsfMetrics {
+    disk_storage_count: U64Counter,
```
Comment (yukkit) [inline tskv/src/tseries_family.rs:615]: gauge

Comment (ZuoTiJia) [reply tskv/src/tseries_family.rs:615]: already fix

File: common/metrics/src/lib.rs [hunk 28]
```diff
@@ -1,12 +1,85 @@
...
+///
+///fn example()  {
+///    let register = MetricsRegister::default();
+///    let options = U64HistogramOptions::new(vec![1, 10, 20, 30, 40, 50, 60, 70, 80, 90]);
+///    let metric = register.register_metric::<U64Histogram>("example", "example metrics", options);
+///    let histogram = metric.recorder([("database", "test"), ("tenant", "cnosdb")]);
+///    histogram.record(5);
+/// }
+///
+/// ```
```
Comment (yukkit) [inline common/metrics/src/lib.rs:62]: this looks amazing

**Review by yukkit — APPROVED** [review]

**Pull request merged**